### PR TITLE
Remember failures from previous builds

### PR DESCRIPTION
**Pull request by jakemac53**
Further work towards https://github.com/dart-lang/build/issues/910, we will now remember all failures (regardless of --fail-on-severe flag), and if you provide that flag we will fail the build.

The current message is somewhat lacking, looks roughly like this:
```
[SEVERE] Build: Failed after 6.9s
There were 1 actions with `severe` logs and --fail-on-severe was passed
```
I added another action item to https://github.com/dart-lang/build/issues/910 to add a `--rerun-failing-actions` flag to get more information. I don't think its worth trying to store all the information about previous failures as it could get quite large if one failure caused a large number of other cascading failures.

### CODE REVIEW

**Comment by jakemac53** [comment]
@chalin This should at least provide all the functionality you need in order to make sure builds fail always with --fail-on-severe, even if its not entirely clear exactly what caused the failure in the followup build case.

File: build_runner/lib/src/asset_graph/graph.dart [hunk 1]
```diff
@@ -22,6 +22,13 @@ part 'serialization.dart';
 
 /// All the [AssetId]s involved in a build, and all of their outputs.
 class AssetGraph {
+  /// A map of phase number to primary inputs that failed during that phase.
+  ///
+  /// See [markActionFailed] and [markActionSucceeded] for more info.
+  final Map<int, Set<AssetId>> _failedActions;
+  UnmodifiableMapView<int, Iterable<AssetId>> get failedActions =>
```
Comment (natebosch) [inline build_runner/lib/src/asset_graph/graph.dart]: why do we need to expose this as `UnmodifiableMapView` instead of `Map`?

Comment (natebosch) [reply build_runner/lib/src/asset_graph/graph.dart]: I wasn't clear...

It makes sense to me to wrap it in `UnmodifiableMapView` for defensiveness.. I'm not sure why the getter's signature also needs to say that.

Comment (jakemac53) [reply build_runner/lib/src/asset_graph/graph.dart]: Oh it probably doesn't I guess, it makes it a bit more clear that its not modifiable though?

Comment (natebosch) [reply build_runner/lib/src/asset_graph/graph.dart]: The interface is the same, the methods you can't use will still autocomplete in your editor etc. It's only if you hover over it and look for the type that you'd notice.

Comment (jakemac53) [reply build_runner/lib/src/asset_graph/graph.dart]: > It's only if you hover over it and look for the type that you'd notice.

Ya but that is a fairly common use case right? At least that is usually how I figure out the type of a thing.

Why wouldn't we want to expose this type?

Comment (natebosch) [reply build_runner/lib/src/asset_graph/graph.dart]: 🤷‍♂️ 

I'm fine with it if you think it adds value.

**Pull request merged**